hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/ec/reconstruction/ECReconstructionCoordinator.java

```diff
@@ -0,0 +1,417 @@
+    }
+
+  }
+
+  void reconstructECBlockGroup(BlockLocationInfo blockLocationInfo,
+      ECReplicationConfig repConfig,
+      SortedMap<Integer, DatanodeDetails> targetMap)
+      throws IOException {
+    long safeBlockGroupLength = blockLocationInfo.getLength();
+    List<Integer> missingContainerIndexes =
+        targetMap.keySet().stream().collect(Collectors.toList());
+
+    // calculate the real missing block indexes
+    int dataLocs = ECBlockInputStreamProxy
+        .expectedDataLocations(repConfig, safeBlockGroupLength);
+    List<Integer> toReconstructIndexes = new ArrayList<>();
+    for (int i = 0; i < missingContainerIndexes.size(); i++) {
+      Integer index = missingContainerIndexes.get(i);
+      if (index <= dataLocs || index > repConfig.getData()) {
+        toReconstructIndexes.add(index);
+      }
+      // else padded indexes.
+    }
+
+    // Looks like we don't need to reconstruct any missing blocks in this block
+    // group. The reason for this should be block group had only padding blocks
+    // in the missing locations.
+    if (toReconstructIndexes.size() == 0) {
+      if (LOG.isDebugEnabled()) {
+        LOG.debug("Skipping the reconstruction for the block: "
+            + blockLocationInfo.getBlockID() + ". In the missing locations: "
+            + missingContainerIndexes
+            + ", this block group has only padded blocks.");
+      }
+      return;
+    }
+
+    try (ECBlockReconstructedStripeInputStream sis
+        = new ECBlockReconstructedStripeInputStream(
+        repConfig, blockLocationInfo, true,
+        this.containerOperationClient.getXceiverClientManager(), null,
+        this.blockInputStreamFactory, byteBufferPool,
+        this.ecReconstructExecutor)) {
+
+      ECBlockOutputStream[] targetBlockStreams =
+          new ECBlockOutputStream[toReconstructIndexes.size()];
+      ByteBuffer[] bufs = new ByteBuffer[toReconstructIndexes.size()];
+      for (int i = 0; i < toReconstructIndexes.size(); i++) {
+        OzoneClientConfig configuration = new OzoneClientConfig();
+        // TODO: Let's avoid unnecessary bufferPool creation for
+        BufferPool bufferPool =
+            new BufferPool(configuration.getStreamBufferSize(),
+                (int) (configuration.getStreamBufferMaxSize() / configuration
+                    .getStreamBufferSize()),
+                ByteStringConversion.createByteBufferConversion(false));
+        targetBlockStreams[i] =
+            new ECBlockOutputStream(blockLocationInfo.getBlockID(),
+                this.containerOperationClient.getXceiverClientManager(),
+                Pipeline.newBuilder().setId(PipelineID.valueOf(
+                    targetMap.get(toReconstructIndexes.get(i)).getUuid()))
+                    .setReplicationConfig(repConfig).setNodes(ImmutableList
+                    .of(targetMap.get(toReconstructIndexes.get(i))))
+                    .setState(Pipeline.PipelineState.CLOSED).build(),
```

Comment:
Nit: please assign `targetMap.get(toReconstructIndexes.get(i))` to local var, and reuse suggested helper method from `ECContainerOperationClient`.

Reply:
done